Implemented dispatch for mimetic inner product selection

francoishamon opened the pull request:
This PR modifies the hybrid/mimetic FVM solver to allow the user to select a mimetic inner product from the input file in the `<FiniteVolume>` block of the XML file.

To do that I adapted some constructs used in `FiniteElementDiscretization`, `FiniteElementDispatch`, and the folder `elementFormulations`. Each mimetic inner product has its own class with a static member function that implements the calculation of the transmissibility matrix. To select the inner product in `SinglePhaseHybridFVM.cpp`, I use a dispatch just before launching the `FluxKernel` (now templated on the inner product type).

Most of the changes of the PR come from the fact that, after the addition of the new template, I was not able to compile with explicit instantiation of the functions in `SinglePhaseHybridFVMKernels.cpp` anymore (maybe there is a workaround but I could not find it). So I moved all the function implementations to the .hpp.

Related to https://github.com/GEOSX/integratedTests/pull/107.

francoishamon commented:
@klevzoff @rrsettgast I am done with the work that I wanted to do in this PR. This refactoring PR is not urgent at all and can wait, but if you have some time over the next weeks it is ready for review.

klevzoff reviewed (approved):
Looks great!

The only issue I have is with the inheritance of `HybridMimeticDiscretization` from `FluxApproximationBase`. Or rather the fact that `FluxApproximationBase` contains the stencils and related methods and `HybridMimeticDiscretization` has to inherit those and implement stubs. Cleaning up that hierarchy probably should be its own PR though.

francoishamon commented:
Yes I agree with you about the class hierarchy in `finiteVolume`. I will do the cleanup in a separate PR to avoid making too many changes in this one. Thanks!

francoishamon commented on src/coreComponents/finiteVolume/MimeticInnerProductDispatch.hpp:
```diff
@@ -0,0 +1,120 @@
+  }
+  else
+  {
+    GEOSX_ERROR( "mimeticInnerProductDispatch() is not implemented for input of "<<LvArray::system::demangleType( &input ) );
```
Done

francoishamon commented on src/coreComponents/finiteVolume/MimeticInnerProductDispatch.hpp:
```diff
@@ -0,0 +1,120 @@
+  }
+  else
+  {
+    GEOSX_ERROR( "mimeticInnerProductDispatch() is not implemented for input of "<<typeid(input).name() );
```
Done

francoishamon commented:
@klevzoff @rrsettgast I modified the `HybridMimeticDiscretization` class, which now inherits directly from `Group` and not from `FluxApproximationBase` anymore. I changed the `FiniteVolumeManager` to reflect the change (thanks @klevzoff for the help!). I also added some checks in the flow solvers to make sure that the selected discretization is compatible with the solver.

francoishamon commented on src/coreComponents/finiteVolume/mimeticInnerProducts/BdVLMInnerProduct.hpp:
```diff
@@ -0,0 +1,217 @@
+  // at this point we have N ( N^T R )^-1 N^T and \tilde{ \lambda }
+  // we have to compute I - R ( R^T R )^-1 R^T and sum
+
+  // 6) compute ( R^T R )^-1
+  real64 temp[ NF ][ 3 ] = {{ 0 }};
+  LvArray::tensorOps::copy< NF, 3 >( temp, cellToFaceMat ); // <<-- TODO: ask about this, we could have Rij_eq_AkiAkj instead
+  LvArray::tensorOps::Rij_eq_AkiBkj< 3, 3, NF >( work_dimByDim,
+                                                 temp,
+                                                 cellToFaceMat );
```
@corbett5 In this function, I need to compute `R^T R`, where `R` is a rectangular matrix of size `NF x 3`. Currently I create a temporary matrix to be able to use `LvArray::tensorOps::Rij_eq_AkiBkj`. Is there already something in LvArray to do the same thing without the temporary matrix? If not, would it make sense to add something like `Rij_eq_AkiAkj`?

corbett5 replied:
I don't think there is currently anything to do this since the `restrict` keyword prevents the usage of `AkiBkj`. But yes given this restriction it would absolutely make sense to add. @rrsettgast and @CusiniM have a PR to do the same thing for other functions.

CusiniM replied:
@francoishamon We can quickly add it to that PR (216) if you want.

francoishamon replied:
Yes that would be great (if you don't mind). Do you want to do it? Or should I do it and commit to your PR? Thanks

CusiniM replied:
I have done it. I just need to add a unit test for it.
